pkg/reconciler/managed/changelogger.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2024 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package managed
+
+import (
+	"context"
+	"fmt"
+
+	changelogs "github.com/crossplane/crossplane-runtime/apis/changelogs/proto/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reference"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// changeLogger processes changes to resources and helps to send them to the
+// change log service
+type changeLogger struct {
+	enabled         bool
+	client          changelogs.ChangeLogServiceClient
+	providerVersion string
+	log             logging.Logger
+	record          event.Recorder
```

Comment:
This is maybe controversial, but I generally prefer to keep logs and events emitted by the code that orchestrates everything else - i.e. the reconcile loop. Especially if I'm just logging/eventing for errors, I think it's better to write dependencies (like this change logger) as more of a library (without loggers etc) and have the caller (the Reconciler) log the errors it returns. This way you get one part of the codebase that's responsible for all logging/eventing etc.

Generally the only exception I'll make to this is if a dependency (like the change logger) has something really interesting/useful to log that _isn't_ an error.

Reply:
Cool cool!! more than happy to follow the existing patterns in the `Reconcile()` code base. Now with the `ChangeLogger` interface in https://github.com/crossplane/crossplane-runtime/pull/754/commits/c23f09b8314e5a7d11e0438243572a0b38cfaff1, an error is returned from `RecordChangeLog()` and `Reconcile()` handles that with debug logging and recording a warning event.